src/vss_tools/vspec/tree.py

```diff
@@ -0,0 +1,541 @@
+# Copyright (c) 2023 Contributors to COVESA
+#
+# This program and the accompanying materials are made available under the
+# terms of the Mozilla Public License 2.0 which is available at
+# https://www.mozilla.org/en-US/MPL/2.0/
+#
+# SPDX-License-Identifier: MPL-2.0
+from __future__ import annotations
+import re
+from typing import Any
+import uuid
+
+from anytree import Node, PreOrderIter, find, findall
+from copy import deepcopy
+
+from vss_tools import log
+from vss_tools.vspec.vspec import deep_update
+from vss_tools.vspec.model import (
+    VSSData,
+    VSSDataDatatype,
+    VSSDataStruct,
+    VSSRaw,
+    get_vss_raw,
+    VSSDataBranch,
+    resolve_vss_raw,
+    ModelValidationException,
+)
+from pydantic import ValidationError
+from vss_tools.vspec.datatypes import Datatypes, dynamic_datatypes
+
+SEPARATOR = "."
+
+
+class NotMergeableException(Exception):
+    pass
+
+
+class NoRootsException(Exception):
+    pass
+
+
+class MultipleRootsException(Exception):
+    pass
+
+
+class InvalidExpansionEntryException(Exception):
+    pass
+
+
+class NoVSSDataException(Exception):
+    pass
+
+
+class VSSNode(Node):  # type: ignore[misc]
+    """
+    Our Anytree node class
+    It contains the actual node data in the `data` attribute
+    """
+
+    separator = SEPARATOR
+
+    def __init__(
+        self, name: str, fqn: str | None, data: dict[str, Any], **kwargs: Any
+    ) -> None:
+        super().__init__(name, **kwargs)
+        self.data = get_vss_raw(data, fqn)
+        self.uuid: str | None = None
+
+    def _post_attach(self, parent: VSSNode):
+        """
+        Updating the data fqn when getting reattached.
+        We need the fqn in the data for validation purposes.
+        """
+        log.debug(
+            f"Got attached to parent='{parent.get_fqn()}', new fqn='{self.get_fqn()}'"
```

Comment:
I do not know how consistent we have been in the past, but at least sometimes we have tried to avoid using f-strings in at least debug/info logging, with the rationale that it may result in extra processing to generate the f-string even if not printed in the end. I.e. better to use `%s`. See link below.

https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/logging-fstring-interpolation.html

When testing with whole VSS catalog I get a gut feeling that generating for example JSON takes quite much longer time compared to before. I have not done any thorough tests. If you have time feel free to compare the result with this PR and previous to see if there are any significant changes.

Reply:
On the other hand, just using `%s` will not help here as maybe the `get_fqn`call is the costly part. If we verify that it is costly then maybe we need a work-around like. (But I do not really know if it actually is a problem from performance perspective)

```
if logger.isEnabledFor(logging.DEBUG):
    logger.debug(f"var1 {var1}, var2 {var2}") 
```

Reply:
I am aware of the f string trade off for logging but usually still use them in favor for readability.
I will make some performance measurements.
Regarding `get_fqn()` I am thinking right now to make it a member and update it when the node gets attached / re-attached

Reply:
Indeed performance is not good. Will investigate. Probably unnecessary deepcopys somewhere

Reply:
Fixed the performance. One performance finding is still there that is not a quick fix though and might need some more work